content/zh/2023-10-29-collect-and-analyse-nginx-accesslog-by-lal/index.md

```diff
@@ -0,0 +1,157 @@
+    return 1, timestamp, newRecord
+end
+```
+
+## OAP 日志分析
+### LAL定义
+在 filter 中，我们通过条件判断，只处理 `service=nginx::nginx` 的服务，其他服务依旧走默认逻辑：
+
+第一步，使用 json 指令对日志进行解析，解析的结果会被存放到 parsed 字段中，通过 parsed 字段我们可以获取 json 日志中的字段信息。
+
+第二步，使用 timestamp 指令解析 parsed.time 并将其赋值给日志的 timestamp 字段，这里的 time 就是access log json 中的 time。
+
+第三步，使用 tag 指令给日志打上对应的标签，标签的值依然可以通过 parsed 字段获取。
+
+第四步，使用 metrics 指令从日志中提取出指标信息，我们共提取了四个指标：
+- `nginx_log_count`：Nginx 每次请求都会生成一条 access log，该指标可以帮助我们统计 Nginx 当前的请求数。
+- `nginx_request_time`：access log 中会记录请求时间，该指标可以帮助我们统计上游接口的响应时长。
+- `nginx_body_bytes_sent`：body 大小指标可以帮助我们了解网关上的流量情况。
+- `nginx_status_code`：状态码指标可以实现对状态码的监控，如果出现异常上涨可以结合 alarm 进行告警。
+
+```yaml
+rules:
+  - name: default
+    layer: GENERAL
+    dsl: |
+      filter {
+        if (log.service == "nginx::nginx") {
+          json {
+            abortOnFailure true
+          }
+    
+          extractor {
+            timestamp parsed.time as String, "yyyy-MM-dd'T'HH:mm:ssXXX"
+            tag status: parsed.status
+            tag remote_addr: parsed.remote_addr
+      
+            metrics {
+              timestamp log.timestamp as Long
+              labels service: log.service, instance: log.serviceInstance
+              name "nginx_log_count"
+              value 1
+            }
+            metrics {
+              timestamp log.timestamp as Long
+              labels service: log.service, instance: log.serviceInstance
+              name "nginx_request_time"
+              value parsed.request_time as Double
+            }
+            metrics {
+              timestamp log.timestamp as Long
+              labels service: log.service, instance: log.serviceInstance
+              name "nginx_body_bytes_sent"
+              value parsed.body_bytes_sent as Long
+            }
+            metrics {
+              timestamp log.timestamp as Long
+              labels service: log.service, instance: log.serviceInstance, status: parsed.status
+              name "nginx_status_code"
+              value 1
+            }
+          }
+        }
+      
+        sink {
+        }
+      }
+```
+经过 LAL 处理后，我们已经可以在日志面板看到日志信息了，接下来我们将对 LAL 中提取的指标进行进一步分析：
+
+![nginx-log](./nginx-log.jpg)
+
+### MAL定义
+在 MAL 中，我们可以对上一步 LAL 中提取的指标进行进一步的分析聚合，下面的例子里：
+
+nginx_log_count、nginx_request_time、nginx_status_code 使用 sum 聚合函数处理，并使用 SUM 方式 downsampling，
+
+nginx_request_time 使用 avg 聚合函数求平均值，默认使用 AVG 方式 downsampling。
+
+完成聚合分析后，SkyWalking Meter System 会完成对上述指标的持久化。
+
+```yaml
+expSuffix: service(['service'], Layer.GENERAL)
+metricPrefix: nginx
+metricsRules:
+  - name: cpm
+    exp: nginx_log_count.sum(['service']).downsampling(SUM)
+  - name: avg_request_time
+    exp: nginx_request_time.avg(['service'])
+  - name: body_bytes_sent_count
+    exp: nginx_body_bytes_sent.sum(['service']).downsampling(SUM)
+  - name: status_code_count
+    exp: nginx_status_code.sum(['service','status']).downsampling(SUM)
+```
+最后，我们便可以来到 SkyWalking UI 页面新建 Nginx 仪表板，使用刚刚 MAL 中定义的指标信息创建 Nginx Dashboard：
```

Comment:
It'd be very helpful if you can also share the sample dashboard json file so that readers can import the dashboard directly

Reply:
Or a screenshot?

We planned to provide official Nginx dashboard, but no one did yet.

Are you interested in pushing this to the upstream?

Reply:
> It'd be very helpful if you can also share the sample dashboard json file so that readers can import the dashboard directly

The dashbord json file is very long, I wil upload it to repo I menthioned in blog.

Reply:
> Or a screenshot?
> 
> We planned to provide official Nginx dashboard, but no one did yet.
> 
> Are you interested in pushing this to the upstream?

Yes, I am glad to do this.